File: tianshu/interfaces/web/mock_llm.py
from __future__ import annotations

import json
import re

from fastapi import FastAPI, Request
from fastapi.responses import JSONResponse


def build_mock_app() -> FastAPI:
    """本地 OpenAI 兼容 mock 服务,用于无 key 端到端验证全流程。"""

    app = FastAPI(title="tianshu-mock-llm")

    @app.get("/healthz")
    async def healthz():
        return {"status": "ok"}

    @app.post("/v1/chat/completions")
    async def chat(req: Request):
        body = await req.json()
        messages = body.get("messages", [])
        tools = body.get("tools", [])
        last_user = ""
        for m in reversed(messages):
            if m.get("role") in ("user", "tool") and isinstance(m.get("content"), str):
                last_user = m["content"]
                break

        tool_names = [t["function"]["name"] for t in tools]
        content = last_user
        has_user_msg = any(m.get("role") == "user" for m in messages)

        if tools and has_user_msg:
            if not _contains_tool_result(messages):
                first = tool_names[0]
                fn = _pick_tool_fn(first, last_user)
                return _resp(
                    content=None,
                    tool_calls=[
                        {
                            "id": "call_mock_1",
                            "type": "function",
                            "function": {"name": first, "arguments": json.dumps(fn, ensure_ascii=False)},
                        }
                    ],
                )

        return _resp(content=f"[mock 回复] 收到: {content[:100]}")

    return app


def _contains_tool_result(messages: list[dict]) -> bool:
    return any(m.get("role") == "tool" for m in messages)


def _pick_tool_fn(name: str, last_user: str) -> dict:
    if name == "call_agent":
        return {"agent": "assistant", "task": last_user}
    if name == "fetch_url":
        m = re.search(r"https?://\S+", last_user)
        return {"url": m.group(0) if m else "https://example.com"}
    if name in ("read_file", "write_file"):
        if name == "read_file":
            return {"path": "workspace/test.txt"}
        return {"path": "workspace/test.txt", "content": last_user}
    if name == "run_shell":
        return {"command": "echo mock"}
    if name == "load_skill":
        return {"name": "chat"}
    if name == "produce_plan":
        return {
            "subtasks": [
                {"worker": "assistant", "goal": last_user},
            ]
        }
    if name == "list_dir":
        return {}
    return {}


def _resp(content: str | None, tool_calls: list | None = None) -> JSONResponse:
    msg: dict = {}
    if content is not None:
        msg["content"] = content
    if tool_calls:
        msg["tool_calls"] = tool_calls
    return JSONResponse(
        {
            "id": "chatcmpl-mock",
            "object": "chat.completion",
            "choices": [{"index": 0, "message": msg, "finish_reason": "stop"}],
            "usage": {
                "prompt_tokens": 100,
                "completion_tokens": 1,
                "total_tokens": 101,
                "prompt_tokens_details": {"cached_tokens": 60},
                "prompt_cache_hit_tokens": 60,
                "prompt_cache_miss_tokens": 40,
            },
        }
    )


app = build_mock_app()

if __name__ == "__main__":
    import uvicorn

    uvicorn.run(app, host="127.0.0.1", port=9100)

File: tianshu-ts/src/interfaces/web/mock_llm.ts
import { createServer } from "node:http";
import type { Server } from "node:http";
import { jsonRes, readJson } from "./ws.js";

function pickToolFn(name: string, lastUser: string): Record<string, unknown> {
  if (name === "call_agent") return { agent: "assistant", task: lastUser };
  if (name === "fetch_url") {
    const m = /https?:\/\/\S+/.exec(lastUser);
    return { url: m ? m[0] : "https://example.com" };
  }
  if (name === "read_file") return { path: "workspace/test.txt" };
  if (name === "write_file") return { path: "workspace/test.txt", content: lastUser };
  if (name === "run_shell") return { command: "echo mock" };
  if (name === "load_skill") return { name: "chat" };
  if (name === "produce_plan") return { subtasks: [{ worker: "assistant", goal: lastUser }] };
  if (name === "list_dir") return {};
  return {};
}

function containsToolResult(messages: Array<Record<string, any>>): boolean {
  return messages.some((m) => m.role === "tool");
}

function response(content: string | null, toolCalls: unknown[] | null = null, reasoning: string | null = null): Record<string, unknown> {
  const msg: Record<string, unknown> = {};
  if (content !== null) msg.content = content;
  if (toolCalls) msg.tool_calls = toolCalls;
  if (reasoning) msg.reasoning_content = reasoning;
  return {
    id: "chatcmpl-mock",
    object: "chat.completion",
    choices: [{ index: 0, message: msg, finish_reason: "stop" }],
    usage: {
      prompt_tokens: 100,
      completion_tokens: 1,
      total_tokens: 101,
      prompt_tokens_details: { cached_tokens: 60 },
      prompt_cache_hit_tokens: 60,
      prompt_cache_miss_tokens: 40,
    },
  };
}

export function createMockServer(host = "127.0.0.1", port = 9100): Server {
  return createServer(async (req, res) => {
    const url = new URL(req.url ?? "/", "http://localhost");
    if (req.method === "GET" && url.pathname === "/healthz") {
      return jsonRes(res, 200, { status: "ok" });
    }
    if (req.method === "POST" && url.pathname === "/v1/chat/completions") {
      const body = await readJson(req);
      const messages: Array<Record<string, any>> = body.messages ?? [];
      const tools: Array<Record<string, any>> = body.tools ?? [];
      let lastUser = "";
      for (const m of [...messages].reverse()) {
        if ((m.role === "user" || m.role === "tool") && typeof m.content === "string") {
          lastUser = m.content;
          break;
        }
      }
      const toolNames = tools.map((t) => t.function?.name).filter(Boolean);
      const hasUserMsg = messages.some((m) => m.role === "user");
      if (toolNames.length && hasUserMsg && !containsToolResult(messages)) {
        const first = toolNames[0];
        const fn = pickToolFn(String(first), lastUser);
        return jsonRes(
          res,
          200,
          response(null, [
            { id: "call_mock_1", type: "function", function: { name: String(first), arguments: JSON.stringify(fn) } },
          ], `用户请求「${lastUser.slice(0, 60)}」,需要获取更多信息,我选择调用工具 ${first} 来完成任务。`),
        );
      }
      return jsonRes(res, 200, response(`[mock 回复] 收到: ${lastUser.slice(0, 100)}`, null, `用户请求「${lastUser.slice(0, 60)}」,信息已足够,我直接整理回答。`));
    }
    return jsonRes(res, 404, { ok: false, error: "404 Not Found" });
  });
}